lightning/restore/restore.go

```diff
@@ -1742,45 +1829,54 @@ func (cr *chunkRestore) deliverLoop(
 			}
 		}
 
-		// Write KVs into the engine
-		start := time.Now()
+		// in local mode, we should write these checkpoint after engine flushed
+		// we are allowed to save checkpoint when the disk quota state moved to "importing"
+		// since all engines are flushed.
+		canSaveCheckpoint := !rc.isLocalBackend() || atomic.LoadInt32(&rc.diskQuotaState) == diskQuotaStateImporting
 
-		if err = dataEngine.WriteRows(ctx, columns, dataKVs); err != nil {
-			deliverLogger.Error("write to data engine failed", log.ShortError(err))
-			return
-		}
-		if err = indexEngine.WriteRows(ctx, columns, indexKVs); err != nil {
-			deliverLogger.Error("write to index engine failed", log.ShortError(err))
-			return
-		}
+		func() {
+			rc.diskQuotaLock.RLock()
```

Comment:
in #446 
> we block the write to the **largest engines** until the remaining total is less than the quota

but here we block all engines. is this expected?

Reply:
Shall we also block `encodeLoop` ? If `deliverLoop` blocks while `encodeLoop` continues processing data, the data may stay in channel and may cause "out of memory".

Reply:
the channel is limit to `maxKVQueueSize` capacity and so `encodeLoop` will be blocked after full 🤔

Reply:
If `deliverLoop` was blocked in "#1923"(`rc.diskQuotaLock.RLock()`), it would not save checkpoint, would it?

Reply:
yes, the current implementation can not handle this if the previous check isImporting is false and the  disk-quote loop holds the lock and do importing.